### ISSUE TITLE
Hardware peripherals, dynamic class loading

### ISSUE THREAD

**Issue by maxbrito500** [FlintVN/FlintESPJVM#2]
Hello,

Thank you for publishing this project. I think most capable Java developers can write C++ code but it isn't exactly the same when you are used to the tooling and easier code writing of Java. Your project really gives fresh air for ESP32.

Have a few questions before going deeper into the project:

+ How are classes loaded?
+ How can WiFi/Bluetooth be accessed?
+ Memory card access already built-in?
+ Access to built-in displays?

I know this project is still fresh, but what are the plans for loading classes or jar files dynamically from memory cards? Right now a limitation ESP32 is the need to flash the firmware or use some tricks for app loading as seen recently at Flipper. Since this is a JVM, would assume that you've solved this heavy limitation on ESP32. We can now compile the class files and then run them on ESP32 devices with ease.

The second question are the basic peripherals. If we can dynamically load classes from memory cards, then connection to WiFi, bluetooth and basic things like controling the pixels/texts on displays comes next.

How can we help add these functionalities?

Thank you!

**Comment by ElectricThanhTung** [comment]
Hi @maxbrito500 
- How are classes loaded.
Currently, you can use vscode with FlintJVM Debug extension to load classes to ESP32. It will automatically load all class files in the specified folder into ESP32 when the "install" parameter is set to true in the launch.json file. 
![image](https://github.com/user-attachments/assets/14027b69-6e26-4f9f-9c38-47a3119a8bbc)

This method can be used with all ESP32 chips. But if your ESP32 supports USB communication like ESP32-S2 or ESP32-S3. It is simpler, you just need to press the boot button on the board then it will switch to USB storage mode (wait about 2s). Press again if you want to return to run mode.
![image](https://github.com/user-attachments/assets/7ec1b6e7-b282-4ed7-8246-6dc33a9d1329)

And I'm also developing a tool to support file browsing via UART. I'll let you know when it's up and running.
![image](https://github.com/user-attachments/assets/b3932bf0-1f3f-4000-961d-43cc5aae801e)

**Comment by maxbrito500** [comment]
This is looking so good!! :-)

So basically just drag and drop new classes instead a very long compilation toolchain and need to upload a new firmware for every little change. Plus we get the debug environment and easy code programming as Java developers enjoy on the IDE.

What about Wi-Fi, bluetooth and memory card integration, I imagine there is way to incorporate C libraries into java libraries?

**Comment by ElectricThanhTung** [comment]
And about:
- How can WiFi/Bluetooth be accessed?
- Memory card access already built-in?
- Access to built-in displays
I haven't had time to do all of them yet because I'm alone. But I'll try to complete it in the future.
Or you can also add your own native methods to support them.
See FlintESPJVM/ESPJVM/User/Src/esp_system_native_pin.cpp and FlintESPJVM/ESPJVM/User/Src/esp_system_native.cpp for how to create a native method

**Comment by ElectricThanhTung** [comment]
Hi @maxbrito500 
"What about Wi-Fi, bluetooth and memory card integration, I imagine there is way to incorporate C libraries into java libraries?"
I think I will do that.

**Comment by maxbrito500** [comment]
Thank you for the invitation. I'm not an expert on C but will gladly be around here to help writing code.

Will just ask a lot of questions until I can learn and understand better, hope you don't mind. In return, I will write these notes as documentation to help others get started. We don't have a forum here, so I will open new issues with the questions if that is OK.

**Comment by hz2shz** [comment]
but on macos, after i press boot button on esp32, i can not say any usb storage locaiton in finder.

**Comment by hz2shz** [comment]
maybe i need to purchuse an windows computer

**Comment by ElectricThanhTung** [comment]
> but on macos, after i press boot button on esp32, i can not say any usb storage locaiton in finder.

No. usb storage mode only available in ESP32-S2 and ESP32-S3 with USB port not UART port.
